### PR TITLE
Use directly OPENSSL_zalloc in OPENSSL_sk_dup and OPENSSL_sk_deep_copy

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Directly allocate sk_TYPE_ stack  while duplicating them.
Same as #1228 , I delete the source branch by error.

### CODE REVIEW

File: crypto/stack/stack.c [hunk 1]
```diff
@@ -36,18 +36,14 @@ OPENSSL_sk_compfunc OPENSSL_sk_set_cmp_func(OPENSSL_STACK *sk, OPENSSL_sk_compfu
...
-    s = OPENSSL_realloc((char *)ret->data,
-                        (unsigned int)sizeof(char *) * sk->num_alloc);
-    if (s == NULL)
+    if ((ret = OPENSSL_zalloc(sizeof(*ret))) == NULL)
+        return NULL;
+    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * sk->num_alloc)) == NULL)
         goto err;
-    ret->data = s;
 
     ret->num = sk->num;
```
Comment (richsalz) [inline crypto/stack/stack.c]: we could replace all those "ret-> = sk->" lines with a single *ret=*sk and then the memcpy

File: crypto/stack/stack.c [hunk 2]
```diff
@@ -60,25 +56,24 @@ OPENSSL_STACK *OPENSSL_sk_dup(OPENSSL_STACK *sk)
...
+                             OPENSSL_sk_copyfunc copy_func,
                              OPENSSL_sk_freefunc free_func)
 {
     OPENSSL_STACK *ret;
     int i;
 
     if ((ret = OPENSSL_malloc(sizeof(*ret))) == NULL)
-        return ret;
+        return NULL;
     ret->comp = sk->comp;
```
Comment (richsalz) [inline crypto/stack/stack.c]: same structure assignment comment.

File: crypto/stack/stack.c (hunk 1 above)
Comment (FdaSilvaYY) [inline crypto/stack/stack.c]: *ret = *sk 
No ?

**Comment by richsalz** [comment]
+1

**Comment by FdaSilvaYY** [comment]
I update the related pod file too. ;)

**Comment by bryongloden** [comment]
<3

**Comment by richsalz** [comment]
@bryongloden are you going to give positive emoji on everything we do? :)

**Comment by bryongloden** [comment]
Greetings @richsalz! Nope. Problem?

**Comment by richsalz** [comment]
Was just curious.

File: crypto/stack/stack.c [hunk 3]
```diff
@@ -36,49 +36,44 @@ OPENSSL_sk_compfunc OPENSSL_sk_set_cmp_func(OPENSSL_STACK *sk, OPENSSL_sk_compfu
...
-    return (ret);
+    return ret;
  err:
     OPENSSL_sk_free(ret);
-    return (NULL);
+    return NULL;
 }
 
-OPENSSL_STACK *OPENSSL_sk_deep_copy(OPENSSL_STACK *sk, OPENSSL_sk_copyfunc copy_func,
+OPENSSL_STACK *OPENSSL_sk_deep_copy(const OPENSSL_STACK *sk,
```
Comment (mattcaswell) [inline crypto/stack/stack.c]: Whitespace on the end of this line

File: crypto/stack/stack.c [hunk 4]
```diff
@@ -36,49 +36,44 @@ OPENSSL_sk_compfunc OPENSSL_sk_set_cmp_func(OPENSSL_STACK *sk, OPENSSL_sk_compfu
...
-    s = OPENSSL_realloc((char *)ret->data,
-                        (unsigned int)sizeof(char *) * sk->num_alloc);
-    if (s == NULL)
-        goto err;
-    ret->data = s;
+    if ((ret = OPENSSL_malloc(sizeof(*ret))) == NULL)
+        return NULL;
+
+    /* direct structure assignment */
+    *ret = *sk
```
Comment (mattcaswell) [inline crypto/stack/stack.c]: Missing ";"

File: crypto/stack/stack.c [hunk 5]
```diff
@@ -36,49 +36,44 @@ OPENSSL_sk_compfunc OPENSSL_sk_set_cmp_func(OPENSSL_STACK *sk, OPENSSL_sk_compfu
...
 
     if ((ret = OPENSSL_malloc(sizeof(*ret))) == NULL)
-        return ret;
-    ret->comp = sk->comp;
-    ret->sorted = sk->sorted;
-    ret->num = sk->num;
+        return NULL;
+
+    /* direct structure assignment */
+    *ret = *sk
```
Comment (mattcaswell) [inline crypto/stack/stack.c]: Missing ";"

File: crypto/stack/stack.c [hunk 6]
```diff
@@ -36,49 +36,44 @@ OPENSSL_sk_compfunc OPENSSL_sk_set_cmp_func(OPENSSL_STACK *sk, OPENSSL_sk_compfu
...
-        goto err;
-    ret->data = s;
+    if ((ret = OPENSSL_malloc(sizeof(*ret))) == NULL)
+        return NULL;
+
+    /* direct structure assignment */
+    *ret = *sk
 
-    ret->num = sk->num;
+    if ((ret->data = OPENSSL_malloc(sizeof(*ret->data) * sk->num_alloc)) == NULL)
```
Comment (mattcaswell) [inline crypto/stack/stack.c]: Line length (just) exceeds 80 (as defined in coding style)

**Comment by mattcaswell** [comment]
@FdaSilvaYY you did try to compile this right? ;-)

**Comment by kroeckx** [comment]
Isn't that what the CI is for, to compile it for you? :)

**Comment by richsalz** [comment]
need another reviewer and then i'll merge.

**Comment by levitte** [comment]
Looks good.  I just want to see how Travis does with this before setting my mark.
(btw, is it just me, or is Travis very slow these days?)

**Comment by levitte** [comment]
+1

**Comment by richsalz** [comment]
thanks!  merged.

**Pull request closed without merging**